jar.go

```diff
@@ -295,6 +299,101 @@ func (j *Jar) cookies(u *url.URL, now time.Time) (cookies []*http.Cookie) {
 	return cookies
 }
 
+// AllCookies returns all cookies for the given URL, or a subpath of the given URL.
+// The returned cookies will have Domain, Expires, HttpOnly, Name,
+// Secure, Path, and Value filled out.
+//
+// It returns an empty slice if the URL's scheme is not HTTP or HTTPS.
+func (j *Jar) AllCookies(u *url.URL) (cookies []*http.Cookie) {
+	return j.allCookies(u, time.Now())
+}
+
+// allCookies is like AllCookies but takes the current time as a parameter.
+func (j *Jar) allCookies(u *url.URL, now time.Time) (cookies []*http.Cookie) {
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return cookies
+	}
+	host, err := canonicalHost(u.Host)
+	if err != nil {
+		return cookies
+	}
+	key := jarKey(host, j.psList)
+
+	j.mu.Lock()
+	defer j.mu.Unlock()
+
+	submap := j.entries[key]
+	if submap == nil {
+		return cookies
+	}
+
+	path := u.Path
+	if path == "" {
+		path = "/"
+	}
+	if len(path) > 1 && strings.HasSuffix(path, "/") {
+		path = strings.TrimSuffix(path, "/")
+	}
+
+	var selected []entry
```

Comment:
If it was me I'd put everything from var selected []entry up to and including the sort into a separate function. That seems to break up the doing logic from the "getting all the hosts and paths in the right format" logic